### DISCUSSION TITLE
"$" finds the element but "$$" returns an empty array

### DISCUSSION THREAD

**Discussion by CarlHat-Core** [webdriverio/webdriverio#13662] [Q&A]
Hi,

I am struggling with this piece of code. When I used '$()' the element is found correctly, but using '$$()' with the same query I kept getting an empty array back. I got a bunch of 'Dropdowns: 0' in the console by running the following code. This doesn't make sense to me, since the element was already found displayed.
```
  const dropdown = $(query);
  const dropdowns = $$(query);
  await dopdown.waitForDisplayed({
      timeout: 9000,
      timeoutMsg: "Dropdown not displayed."
  });
  // Wait for the dropdowns
  await browser.waitUntil(async () => {
      console.log(`Dropdowns: ${(await dropdowns).length}`);
      return (await dropdowns).length > index;
  }, {
      timeout: 9000,
      timeoutMsg: "Dropdowns not displayed."
  });
```
If I put a `browser.debug()` right before this, and continue after the breakpoint is hit, everything works properly. Is there something that I am doing wrong here?

**Comment by christian-bromann** [comment]
@CarlHat-Core Hey 👋  thanks for raising this, would you mind sharing an example that we could run ourselves for testing?

**Reply by CarlHat-Core** [reply]
```
const query = 'i.pi.pi-chevron-down.py-2';
const trackedItemDropdown = $(query);
const trackedItemDropdowns = $$(query);
await trackedItemDropdown.waitForDisplayed({
    timeout: MaterialRequest.timeout,
    timeoutMsg: "Tracked item dropdown not displayed."
});
// Wait for the dropdowns to be displayed
await browser.waitUntil(async () => {
    console.log(`Dropdowns: ${(await trackedItemDropdowns).length}`);
    return (await trackedItemDropdowns).length > index;
}, {
    timeout: MaterialRequest.timeout,
    timeoutMsg: "Tracked item dropdowns not displayed."
});
```
This is the selector that I am using. I am not sure whether it would be reproducible on your end, since the code works fine in one environment but not the other. The ui should be the same in both environment. I don't know what is making it work in one but not the other. I checked in dev tool, and the selector works fine in both (The `waitForDisplayed` always passing supports this too).

This is the console output for the test environment run. (The one that it fails in)
![image](https://github.com/user-attachments/assets/7f90dd22-c638-4683-870b-0589125903bc)

This is the element in dev tool
![image](https://github.com/user-attachments/assets/cbe06b9e-ff71-412a-8101-e4f9c8519765)

**Comment by troyyer** [comment]
If it's working on one environment and not the other it is potentially a timing issue as you have defined the webelements $$ selector earlier in the code and it may not get any elements from the page html at the time of defining it so if you define it it close to the waituntil command it could work
```
const query = 'i.pi.pi-chevron-down.py-2';
const trackedItemDropdown = $(query);
await trackedItemDropdown.waitForDisplayed({
    timeout: MaterialRequest.timeout,
    timeoutMsg: "Tracked item dropdown not displayed."
});
// Wait for the dropdowns to be displayed
const trackedItemDropdowns = $$(query); //moved code here
await browser.waitUntil(async () => {
    console.log(`Dropdowns: ${(await trackedItemDropdowns).length}`);
    return (await trackedItemDropdowns).length > index;
}, {
    timeout: MaterialRequest.timeout,
    timeoutMsg: "Tracked item dropdowns not displayed."
});
```
this answer from @christian-bromann on a similar question I had raised could be related to your issue
https://github.com/webdriverio/webdriverio/discussions/12001#discussioncomment-8125341
and 
https://github.com/webdriverio/webdriverio/discussions/12001#discussioncomment-8126796

also maybe also use await in the getter statement can help here?
`const trackedItemDropdowns = await $$(query);`
